### DISCUSSION TITLE
USB keyboard timing issue?

### DISCUSSION THREAD

**Discussion by dhansel** [hathach/tinyusb#1456] [Q&A]
I wasn't sure whether I should report this as a bug or not so I'm starting here. Please let me know if I should file a proper bug report.

I have the following setup:
- Raspberry Pi pico, overclocked to 250MHz (for DVI output)
- TinyUSB latest version from master
- TinyUSB configured as host to support USB keyboard input

I have tried several keyboards and all work without problems except one, a Logitech K270.

At log level 0 or 1, the keyboard does not work. If log level is 1 then I get the following output when plugging in the keyboard: process_set_config 427: ASSERT FAILED

At log level 2 or 3 the keyboard works consistently without problems.

I figured that this may be a timing problem (i.e. logging helps by slowing things down) and found that if I insert a 15ms wait ("osal_task_delay(15);") in line 556 of src/host/usbh.c, right before the "if(xfer->complete_cb)" statement then the keyboard also works when  logging is disabled. That does seem to confirm that this is a timing issue.

Does anybody know if there is a way to deal with this that does not involve local changes to the TinyUSB code?

I'll be happy to provide more logs but of course if I enable log level 2 then the problem does not occur anymore and log level 1 only produces the single ASSERT_FAILED line.

Thanks!
David

**Comment by dhansel** [comment]
I looked a bit more into this. Capturing the USB traffic between the Pico and the keyboard shows that the the Logitech K270 keyboard responds with a STALL if the host requests the keyboard report descriptor too soon (<10ms?) after switching to the BOOT protocol.

That brings up a question: is there a specific reason why the HID code requests the keyboard report descriptor? The code forces the boot protocol and the boot protocol always uses the same default report format. Maybe I'm missing something but I don't see why it is necessary to request the report descriptor at all.

**Reply by hathach** [reply]
> That brings up a question: is there a specific reason why the HID code requests the keyboard report descriptor? The code forces the boot protocol and the boot protocol always uses the same default report format. Maybe I'm missing something but I don't see why it is necessary to request the report descriptor at all.

this does make sense, though even boot device has its own boot descriptor and device is still required to respond to get descriptor request. We will see why it stalls

**Reply by dhansel** [reply]
I get that the keyboard is still required to respond - and the K270 does, it just appears to need some time after switching to the boot protocol before it can respond to the request.

As far as I understand the point of the boot protocol is to enable simpler implementations of the HID protocol that don't require requesting or parsing the report descriptor. See for example [here](https://www.devever.net/~hl/usbnkro) (section "How USB keyboards work"):

> Since the HID Report Descriptor format is complex, it was felt that BIOS writers might be neither able or inclined to incorporate a full implementation of HID which can parse and understand arbitrary Report Descriptors. Thus, the USB HID specification includes an additional notion of a “boot keyboard”, which can be used by the BIOS when e.g. configuring your PC's BIOS settings.
> A keyboard complying with the “boot protocol” is required to use a specific format for its Input Reports set out in the HID specification. That way, BIOSes can be hardcoded to use and expect that format of Input Report and never even have to look at the corresponding Report Descriptor.

Since the boot protocol report format is fixed, there should be no need to request/parse it. The advantage of skipping that step is that it would simplify the code and remove possible points of failure (like in my case).

I don't see any disadvantages of removing that step. But I'm not an expert on this so maybe I am missing something.

**Comment by hathach** [comment]
1. since the github version changes often so the line number, could you paste the permanent link to the assert here (right click on the line in the github file viewer) here 
2. Also please attach the txt containing the configuration descriptor along with hid descriptor of the K270 kebyoard. If you don't know how to get it, I could help if you specify your pc OS.
3. although at level 2, the issue does not appear, please attach the txt log as well for reference. It help to understand the data flow with K270, which currently only you possess.

**Reply by dhansel** [reply]
1. This is the assert that fails:
    https://github.com/hathach/tinyusb/blob/6ec5174d8b24d1f01a443c74a0a3dbee00523efc/src/class/hid/hid_host.c#L427
    From my USB traffic captures I can tell that the keyboard responds with a STALL after this call:
    https://github.com/hathach/tinyusb/blob/6ec5174d8b24d1f01a443c74a0a3dbee00523efc/src/class/hid/hid_host.c#L463
    I can prevent the stall by inserting a 15ms delay before the tuh_descriptor_get_hid_report call.
2. I'm using Windows 10 - please let me know how to retrieve the descriptor .txt files.
3. Log file is attached.
[usbhidlog.txt](https://github.com/hathach/tinyusb/files/8631111/usbhidlog.txt)

**Reply by hathach** [reply]
thanks for the info, config and hid descriptor can be retrieve using https://www.uwe-sieber.de/usbtreeview_e.html

**Reply by dhansel** [reply]
The USBTreeView output for the keyboard is attached. Note that the K270 is a wireless keyboard that uses the Logitech unifying receiver dongle. The same dongle can support a keyboard and a mouse at the same time.
[descriptors.txt](https://github.com/hathach/tinyusb/files/8631385/descriptors.txt)

**Reply by hathach** [reply]
yes, I am familiar with Logitech unifying dongle. I have an logitech mouse as well. Yours has 3 Hid interface, do you know on which interface the assert occur ? if not, you can slightly modify the code to also output the interface id/number as well.

**Reply by dhansel** [reply]
I can modify the code. Let me know what information you need to identify the interface.

**Reply by hathach** [reply]
change the code to following, it should print out the error when running with log level = 1

```C
static void process_set_config(tuh_xfer_t* xfer)
{
  uint8_t const itf_num = (uint8_t) tu_le16toh(xfer->setup->wIndex);

  // Stall is a valid response for SET_IDLE, therefore we could ignore its result
  if ( xfer->setup->bRequest != HID_REQ_CONTROL_SET_IDLE )
  {
    TU_LOG1_INT(itf_num);
    TU_ASSERT(xfer->result == XFER_RESULT_SUCCESS, );
  }

  uintptr_t const state = xfer->user_data;
  uint8_t const daddr   = xfer->daddr;
```

**Reply by dhansel** [reply]
Output is:
itf_num = 0
process_set_config 431: ASSERT FAILED

**Reply by dhansel** [reply]
I have attached the USB captures for the two cases (in Sigrok PulseView format):
usbhid_stall.sr: Shows the failing USB traffic when plugging in the keyboard until the stall
usbhid_ok.sr: Shows the successful USB traffic if I insert a 1ms delay before sending the report request
(It turns out that even a 1ms delay is sufficient)
[capture.zip](https://github.com/hathach/tinyusb/files/8632076/capture.zip)

**Reply by hathach** [reply]
thanks for the capture, I am not familiar with Sigrok PulseView, can you give a little help for how to open it e.g which software, and which step to import/open it.

**Reply by dhansel** [reply]
You can download PulseView here: http://sigrok.org/wiki/Downloads
It's free/open source. I use the Windows version but I assume Linux/Mac will look the same.
I'm using the "nightly build" version which they say is recommended.

After installing PulseView and extracting the ZIP archive with my captures into a folder
you should be able to just double-click on the ".sr" files to open them in PulseView.
If that doesn't work, start PulseView, click the "New Session" button in the toolbar (top left) and
then within the session toolbar click the "Open" button to load a .sr file.

That should show the raw capture with various levels of USB packet decoding beneath it.
You can use the mouse wheel to zoom in/out and drag with the mouse to move around
within the capture.

**Reply by hathach** [reply]
thanks for detail info, I think we got enough info for analyzing. Though I need to wrap up other works first before looking more closer to this. Please be patient in the mean time.
